pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -321,38 +320,28 @@ func (am *alertmanager) aggregateInhibitMatchers(rules []config.InhibitRule, amu
 // It is not safe to call concurrently.
 func (am *alertmanager) applyConfig(cfg *apimodels.PostableUserConfig) (bool, error) {
 	// First, let's make sure this config is not already loaded
-	var amConfigChanged bool
-	rawConfig, err := json.Marshal(cfg.AlertmanagerConfig)
+	rawConfig, err := json.Marshal(cfg)
 	if err != nil {
 		// In theory, this should never happen.
 		return false, err
 	}
 
-	if am.Base.ConfigHash() != md5.Sum(rawConfig) {
-		amConfigChanged = true
-	}
-
-	if cfg.TemplateFiles == nil {
-		cfg.TemplateFiles = map[string]string{}
-	}
-	cfg.TemplateFiles["__default__.tmpl"] = alertingTemplates.DefaultTemplateString
-
-	// next, we need to make sure we persist the templates to disk.
-	paths, templatesChanged, err := PersistTemplates(am.logger, cfg, am.Base.WorkingDirectory())
-	if err != nil {
-		return false, err
-	}
-	cfg.AlertmanagerConfig.Templates = paths
-
-	// If neither the configuration nor templates have changed, we've got nothing to do.
-	if !amConfigChanged && !templatesChanged {
-		am.logger.Debug("Neither config nor template have changed, skipping configuration sync.")
+	// If configuration hasn't changed, we've got nothing to do.
+	configHash := md5.Sum(rawConfig)
+	if am.Base.ConfigHash() == configHash {
```

Comment:
Does changing the level at which the hash operates break remote alertmanagers? (@santihernandezc may know more than me here). Will the remote alertmanager still return the hash of the inner, AM config and not this new outer object (PostableUserConfig), it'll not match up and reload needlessly - is this possible?

Do all implementations return the hash at the right level?

Reply:
The hash is computed in Grafana and sent to the remote Alertmanager, when we compare configurations we do it using that same hash we sent.
If all config hashes are derived from the same struct, then nothing should break.
Plus, the integration tests are passing 👌

Reply:
The way it's currently implemented, this should not affect the remote alertmanagers as this code is internal to the local alertmanager and does not affect what we store in the database.

However, if this somehow changes then we'd likely want all the hashes to be based on `PostableUserConfig` anyways since it contains `TemplateFiles` which is how we'll be passing templates to the remote alertmanager, correct?

I'll let @santihernandezc confirm though.

Reply:
> this should not affect the remote alertmanagers as this code is internal to the local alertmanager and does not affect what we store in the database.

Agree, we don't use the internal Alertmanager's hash for the remote/forked AM, we use what we receive via the MOA.

> if this somehow changes then we'd likely want all the hashes to be based on PostableUserConfig anyways since it contains TemplateFiles which is how we'll be passing templates to the remote alertmanager, correct?

Yep, AFAIU that's the way we currently do it. The hash we send to the remote Alertmanager is the one we get from the database in `ApplyConfig`. This hash is derived from `cmd.AlertmanagerConfiguration` ([permalink](https://github.com/grafana/grafana/blob/2184592174b8ec1ff20381349773fef1533d4b9e/pkg/services/ngalert/store/alertmanager.go#L74C4-L74C15)), which is a marshaled `PostableUserConfig` ([permalink](https://github.com/grafana/grafana/blob/2184592174b8ec1ff20381349773fef1533d4b9e/pkg/services/ngalert/notifier/alertmanager.go#L213-L221)).